internal/jimm/permissions/relations.go

```
@@ -29,11 +36,19 @@ func (j *permissionManager) AddRelation(ctx context.Context, user *openfga.User,
 	if err != nil {
 		return errors.E(err)
 	}
-	err = j.authSvc.AddRelation(ctx, parsedTuples...)
-	if err != nil {
-		return errors.E(op, errors.CodeOpenFGARequestFailed, err)
+	for i := 0; i < len(parsedTuples); i += BATCH_SIZE_OPENFGA {
+		end := i + BATCH_SIZE_OPENFGA
+		if end > len(parsedTuples) {
+			end = len(parsedTuples)
+		}
+		batch := parsedTuples[i:end]
+
+		err = j.authSvc.AddRelation(ctx, batch...)
```

Do we care about partial failure here and eventual retry?

I haven't checked if `AddRelation` is idempotent.

---

We care because if you try to add 1000 things from terraform and you fail midway when you retry you will receive an error because there is a conflict on write.

I'll expand the PR description a bit to give more context

---

In that case, we may need to swallow the error from `AddRelation`/`RemoveRelation` in the case that the state is already what we're trying to make it be.

I suppose we already had that problem in case one of the sent relations was already added/removed.

Should be fine as a potential follow-up.

---

if i've understood correctly you are saying to check the err type and not return it in case it's `writeDuplicate`.
However i believe openfga will return this error if one single tuple in the list is duplicated, so it might be that the state is not what you want but just one single element is duplicated.

---

Exactly, the write of X tuples can fail because of a single duplicate, whether we're trying to send 100 or 10.

The solution would be for the operation to be idempotent. Either openfga does that or we can do that by retrying the call without the offending duplicate.

Either way, even without the batching this was already a potential problem that the caller would have to resolve by manually removing the duplicate anyway. The batching merely makes it more likely to cause duplicates by only executing some of the batches until an error.